### ISSUE TITLE
Getting message: AttributeError: 'socket' object has not attribute 'readline'

### ISSUE THREAD

**Issue by johncblacker** [shawwwn/uMail#9]
Running on uPy 1.13, specifically an openmv port running on a stm32 f765 (looks like a pyboard).  Trying to connect to my normal email server (spectrum email) and am using port 587 and running the "simple" example.  My guess is that the socket is not being created, but I really don't know.  Here's how I have your code setup right now:
 msg = sock.read(200)
        print('msg: ', msg)
        code = int(msg[:3])
        print('code: ', code)
        #sock.settimeout(DEFAULT_TIMEOUT)
        sock.readline()

The print of msg shows:
msg:  b'220 p-impout009.msg.pkvw.co.charter.net cmsmtp ESMTP server ready\r\n'

which seems to me that everything is OK?  But if that were the case I wouldn't be getting the Attribute... message.  I have tried your code on a esp32 board and it worked fine.  This openmv board doesn't have internet, so I'm using a WINC1500 shield to connect to my wifi.  My wifi connects fine and in fact I can run one of their examples that streams a shapshot from a camera via webserver code to my browser and it works fine - does use sockets, BTW.   Both of these boards are using uPy 1.13, but obviously different ports.  Do you have any ideas regarding what I can do to get around my problem?

**Comment by shawwwn** [comment]
What is the line number from which this error occurs?

**Comment by johncblacker** [comment]
So, I removed all my changes to your code and put it back to distro.

The first error I encounter is:

OSError: [Errno 107] ENOTCONN

occurring on line 32: sock.settimeout(DEFAULT_TIMEOUT)

 

Having researched online and found several places suggesting that you cannot settimeout until after the socket is connected.  Rather than move the statement, 

I simply commented it out.  Then the following error occurs:

attributeError: ‘socket’ object has no attribute ‘readline’

occurring on line 37: sock.readline()

 

I can go into repl and import usocket.socket and there is no defined “readline()” function.

 

Here’s output from help(usocket.socket)

object <class 'socket'> is of type type

__del__ -- <function>

close -- <function>

bind -- <function>

listen -- <function>

accept -- <function>

connect -- <function>

send -- <function>

recv -- <function>

write -- <function>

read -- <function>

sendto -- <function>

recvfrom -- <function>

setsockopt -- <function>

settimeout -- <function>

setblocking -- <function>

 

My version of uPy is: 

(sysname='pyboard', nodename='pyboard', release='1.13.0', version='v1.13-r69 on 2021-02-13', machine='OPENMV3 with STM32F765')

 

Now, on the board that works:

MicroPython v1.13-153-g9c0ac7e6a-dirty on 2020-11-07; Camera Module (i2s) with ESP32

bject <class 'socket'> is of type type

  __del__ -- <function>

  close -- <function>

  bind -- <function>

  listen -- <function>

  accept -- <function>

  connect -- <function>

  send -- <function>

  sendall -- <function>

  sendto -- <function>

  recv -- <function>

  recvfrom -- <function>

  setsockopt -- <function>

  settimeout -- <function>

  setblocking -- <function>

  makefile -- <function>

  fileno -- <function>

  read -- <function>

  readinto -- <function>

  readline -- <function>

  write -- <function>

Guess the problem isn’t in your code; but, rather, the version of uPy that I’m using isn’t compatible with your code.

 

Thanks for the response…my bad.

**Comment by shawwwn** [comment]
I currently don't have a uPy-capable board. If you can make umail work under uPy, I will be happy to open a new branch for your code.

**Comment by johncblacker** [comment]
Well, after a lot of playing around (not really playing) I got it to work; however, the program did end with an OSError: -12 which I gather is an 

out of memory issue.  Have to look into fixing that if I can.  Will get back to you and update the issue on github after I fix the OSError and remove

my print statements.  My limited success was only with the simple example – not the ssl example; but, hey, I did get somewhere with it.

 

John Blacker
